## 6-е домашнее задание - Библиотека numpy

**Студент:** Феоктистов Владислав Сергеевич

**Группа:** НПМбд-01-19

1.	Создайте равномерное разбиение интервала от -1.3 до 2.5 на 64 отрезка.

In [2]:
import numpy as np
np.linspace(-1.3, 2.5, 64)

array([-1.3       , -1.23968254, -1.17936508, -1.11904762, -1.05873016,
       -0.9984127 , -0.93809524, -0.87777778, -0.81746032, -0.75714286,
       -0.6968254 , -0.63650794, -0.57619048, -0.51587302, -0.45555556,
       -0.3952381 , -0.33492063, -0.27460317, -0.21428571, -0.15396825,
       -0.09365079, -0.03333333,  0.02698413,  0.08730159,  0.14761905,
        0.20793651,  0.26825397,  0.32857143,  0.38888889,  0.44920635,
        0.50952381,  0.56984127,  0.63015873,  0.69047619,  0.75079365,
        0.81111111,  0.87142857,  0.93174603,  0.99206349,  1.05238095,
        1.11269841,  1.17301587,  1.23333333,  1.29365079,  1.35396825,
        1.41428571,  1.47460317,  1.53492063,  1.5952381 ,  1.65555556,
        1.71587302,  1.77619048,  1.83650794,  1.8968254 ,  1.95714286,
        2.01746032,  2.07777778,  2.13809524,  2.1984127 ,  2.25873016,
        2.31904762,  2.37936508,  2.43968254,  2.5       ])

2.	Сгенерируйте numpy массив длины $3n$, заполненный циклически числами 1, 2, 3, 1, 2, 3, 1....

In [6]:
import numpy as np
n = 10
np.arange(3*n) % 3 + 1

array([1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1,
       2, 3, 1, 2, 3, 1, 2, 3], dtype=int32)

3.	Создайте массив первых 10 нечетных целых чисел.

In [7]:
import numpy as np
np.arange(1, 20, 2)

array([ 1,  3,  5,  7,  9, 11, 13, 15, 17, 19])

4.  Создайте массив нулей размера 10 x 10, а затем создайте в нём "рамку" из единиц по краям.

In [16]:
import numpy as np
Z = np.zeros((10, 10))
Z[::9,:] = 1
Z[1:9,::9] = 1
Z

array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

5.  Создайте массив 8 x 8 с шахматной доской из нулей и единиц.

In [20]:
import numpy as np
Z = np.zeros((8, 8))
Z[1::2,1::2] = 1
Z[::2,::2] = 1
Z

array([[1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.]])

6.  Создайте функцию, которая создает $n×n$ матрицу с $(i,j)$-элементом, равным $i+j$.

In [25]:
def create_matrix(n):
    A = np.arange(n**2)
    return (A // n + A % n).reshape(n, n)

create_matrix(10)

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10],
       [ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
       [ 3,  4,  5,  6,  7,  8,  9, 10, 11, 12],
       [ 4,  5,  6,  7,  8,  9, 10, 11, 12, 13],
       [ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14],
       [ 6,  7,  8,  9, 10, 11, 12, 13, 14, 15],
       [ 7,  8,  9, 10, 11, 12, 13, 14, 15, 16],
       [ 8,  9, 10, 11, 12, 13, 14, 15, 16, 17],
       [ 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]], dtype=int32)

7. Примените функцию dot для перемножения вектор на вектор, матрицу на вектор и матрицу на матрицу.

In [36]:
import numpy as np

V1 = np.arange(1, 5)
V2 = np.arange(5, 9)
print('V1 x V2 =', np.dot(V1, V2))

A = np.arange(1, 13).reshape((3, 4))
B = np.arange(5, 25).reshape((4, 5))
print('A x B =\n', np.dot(A, B))

V1 x V2 = 70
A x B =
 [[150 160 170 180 190]
 [350 376 402 428 454]
 [550 592 634 676 718]]


8. Вычислите cos(x) и sin(x) на интервале [0,1] с шагом $0.05$, а затем объедините оба массива чисел как строки в один массив.

In [59]:
import numpy as np
X = np.linspace(0, 1, int(1/0.05)+1)
C = np.cos(X)
S = np.sin(X)
print('sin =', C, '\ncos =', S)
np.hstack([C, S])

sin = [1.         0.99875026 0.99500417 0.98877108 0.98006658 0.96891242
 0.95533649 0.93937271 0.92106099 0.9004471  0.87758256 0.85252452
 0.82533561 0.7960838  0.76484219 0.73168887 0.69670671 0.65998315
 0.62160997 0.58168309 0.54030231] 
cos = [0.         0.04997917 0.09983342 0.14943813 0.19866933 0.24740396
 0.29552021 0.34289781 0.38941834 0.43496553 0.47942554 0.52268723
 0.56464247 0.60518641 0.64421769 0.68163876 0.71735609 0.75128041
 0.78332691 0.8134155  0.84147098]


array([1.        , 0.99875026, 0.99500417, 0.98877108, 0.98006658,
       0.96891242, 0.95533649, 0.93937271, 0.92106099, 0.9004471 ,
       0.87758256, 0.85252452, 0.82533561, 0.7960838 , 0.76484219,
       0.73168887, 0.69670671, 0.65998315, 0.62160997, 0.58168309,
       0.54030231, 0.        , 0.04997917, 0.09983342, 0.14943813,
       0.19866933, 0.24740396, 0.29552021, 0.34289781, 0.38941834,
       0.43496553, 0.47942554, 0.52268723, 0.56464247, 0.60518641,
       0.64421769, 0.68163876, 0.71735609, 0.75128041, 0.78332691,
       0.8134155 , 0.84147098])

10. Найдите сумму всех элементов, сумму внутри строк, сумму внутри столбцов, а также среднее значение, дисперсию и стандартное отклонение чисел для каждой строки матрицы $A$. (Подобно тому, как sorted имеет необязательный аргумент key=, многие функции Numpy имеют необязательный аргумент axis=)


In [88]:
import numpy as np
A = np.arange(1,13).reshape(3, 4)
sum = np.sum(A)
sum_row = np.sum(A, axis=-1)
sum_col = np.sum(A.transpose(), axis=-1)
aver = np.average(A, axis=-1)
#D = np.sum((A - aver.reshape(aver.shape[0], 1))**2, axis=-1) / A.shape[1]
D = np.var(A, axis=-1, dtype=float)
sigma = np.sqrt(D)
print('sum:', sum)
print('sum by rows:', sum_row)
print('sum by columns:', sum_col)
print('average by rows:', aver)
print('Dispersion by rows:', D)
print('Standard deviation by rows:', sigma)

sum: 78
sum by rows: [10 26 42]
sum by columns: [15 18 21 24]
average by rows: [ 2.5  6.5 10.5]
Dispersion by rows: [1.25 1.25 1.25]
Standard deviation by rows: [1.11803399 1.11803399 1.11803399]


11.  Отнимите от каждого элемента матрицы $A$ среднее по строке и поделите на стандартное отклонение строки.

In [91]:
import numpy as np
A = np.arange(1,13).reshape(3, 4)
aver = np.average(A, axis=-1)
sigma = np.sqrt(np.var(A, axis=-1, dtype=float))
print((A - aver.reshape(len(aver), 1)) / sigma.reshape(len(sigma), 1))

[[-1.34164079 -0.4472136   0.4472136   1.34164079]
 [-1.34164079 -0.4472136   0.4472136   1.34164079]
 [-1.34164079 -0.4472136   0.4472136   1.34164079]]


12. Отсортируйте матрицу $A$ по 3-ому столбцу, т.е. поменяйте местами строки матрицы так, чтобы 3-й столбец оказался отсортированным. Используйте для этого slicing + argsort + indexing.

In [98]:
import numpy as np
A = np.arange(13,1,-1).reshape(3, 4).transpose()
A[A[:, 2].argsort()]

array([[10,  6,  2],
       [11,  7,  3],
       [12,  8,  4],
       [13,  9,  5]])

13. Посчитайте две матрицы: $B = A^TA$ и $C = AA^T$. Что вы можете о них сказать? Какого они размера? В чём их сходство и отличие?

In [101]:
import numpy as np
A = np.arange(1,13).reshape(3, 4)
B = A.transpose() @ A
C = A @ A.transpose()
print('Matrix B:\n', B, '\nShape:', B.shape)
print('Matrix C:\n', C, '\nShape:', C.shape)

# Сходства: обе матрицы квадратные, одинаковая сумма по диагонали
# Различия: размер матрицы B 4x4, а С - 3x3

Matrix B:
 [[107 122 137 152]
 [122 140 158 176]
 [137 158 179 200]
 [152 176 200 224]] 
Shape: (4, 4)
Matrix C:
 [[ 30  70 110]
 [ 70 174 278]
 [110 278 446]] 
Shape: (3, 3)


14. Найдите сумму диагональных элементов матриц $B$ и $C$.

In [109]:
import numpy as np
A = np.arange(1,13).reshape(3, 4)
B = A.transpose() @ A
C = A @ A.transpose()
sum_B = np.sum(np.diagonal(B))
sum_C = np.sum(np.diagonal(C))
print('The sum of the diagonal elements of the matrix B:', sum_B)
print('The sum of the diagonal elements of the matrix C:', sum_C)

The sum of the diagonal elements of the matrix B: 650
The sum of the diagonal elements of the matrix C: 650


15.	Посчитайте детерминант и ранг матриц $A$, $B$ и $C$, посмотрите на спектр (набор собственных значений) матриц $B$ и $C$, какие выводы вы можете сделать?  
(подсказка: используйте готовые функции из библиотеки np.linalg). 


In [4]:
import numpy as np
A = np.arange(1,13).reshape(3, 4)
B = A.transpose() @ A
C = A @ A.transpose()
rank_A = np.linalg.matrix_rank(A)
rank_B = np.linalg.matrix_rank(B)
rank_C = np.linalg.matrix_rank(C)

try:
    det_A = np.linalg.det(A)
except:
    det_A = "не определен"
det_B = np.linalg.det(B) # B - квадратная матрица
det_C = np.linalg.det(C) # C - квадратная матрица 

print('The matrix A:\n', A, '\nrank(A) =', rank_A, '\ndet(A) =', det_A, end='\n\n')
print('The matrix B:\n', B, '\nrank(B) =', rank_B, '\ndet(B) =', det_B, end='\n\n')
print('The matrix C:\n', C, '\nrank(C) =', rank_C, '\ndet(C) =', det_C)

The matrix A:
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]] 
rank(A) = 2 
det(A) = не определен

The matrix B:
 [[107 122 137 152]
 [122 140 158 176]
 [137 158 179 200]
 [152 176 200 224]] 
rank(B) = 2 
det(B) = 0.0

The matrix C:
 [[ 30  70 110]
 [ 70 174 278]
 [110 278 446]] 
rank(C) = 2 
det(C) = 0.0


16. Численно посчитать определённый интеграл  3-мя методами:  
a)	Прямоугольников $\int\limits_a^b {f(x)dx \approx \Delta x\sum\limits_{k = 0}^N {f({x_k})} }$  
b)	Трапеций $\int\limits_a^b {f(x)dx \approx \frac{{\Delta x}}{2}\left( {f({x_0}) + 2\sum\limits_{k = 1}^{N - 1} {f({x_k})}  + f({x_N})} \right)}$  
c)	По правилу Симпсона  $\int\limits_a^b {f(x)dx \approx \frac{{\Delta x}}{3}\left( {f({x_0}) + 2\sum\limits_{k = 1}^{N/2 - 1} {f({x_{2k}})}  + 4\sum\limits_{k = 1}^{N/2} {f({x_{2k - 1}})}  + f({x_N})} \right)}$  


В реализации этих методов цикл for для суммирования использовать нельзя.  
Нужно посчитать значение интеграла с некоторым шагом интегрирования (например начать с 0.1), потом уменьшить его в десять раз и ещё раз посчитать и так до тех пор пока отличие в ответах будет в 5-м знаке после запятой, т.е. чтобы ошибка была меньше 1e-5. Получить ответы для 3-х разных методов и сделать соответствующие выводы.  

Далее необходимо сравнить полученные оценки с аналитическим решением (т.е. формульным, которое вы должны сами посчитать на листочке и вбить формулу-ответ).   

В качестве интеграла взять один из следующих с номером, сгенерированным случайным образом с помощью функции np.random.randint(10), в качестве seed для генератора случайных чисел взять номер своего студенческого билета.

<table width="80%"
       align="center">
    <tr>
        <td> 0) $\int\limits_3^4 {\frac{{{x^2} + 3}}{{x - 2}}dx}$ </td>
        <td> 1) $\int\limits_{ - 2}^{ - 1} {\frac{{x + 1}}{{{x^3} - {x^2}}}dx}$ </td>
        <td> 2) $\int\limits_1^2 {\frac{{{e^{1/{x^2}}}}}{{{x^3}}}dx}$ </td>
        <td> 3) $\int\limits_1^e {\frac{{\cos (\ln x)}}{x}dx}$ </td>
        <td> 4) $\int\limits_1^e {\frac{{dx}}{{x(1 + {{\ln }^2}x)}}}$ </td>
    </tr>                
    <tr>
        <td> 5) $\int\limits_0^{\pi /2} {{{\cos }^3}\alpha \;d\alpha }$ </td>
        <td> 6) $\int\limits_0^{1/3} {c{h^2}3x\;dx}$ </td>
        <td> 7) $\int\limits_2^3 {\frac{{dy}}{{{y^2} - 2y - 8}}dy}$ </td>
        <td> 8) $\int\limits_{3/4}^2 {\frac{{dx}}{{\sqrt {2 + 3x - 2{x^2}} }}}$ </td>
        <td> 9) $\int\limits_0^2 {\frac{{2x - 1}}{{2x + 1}}dx}$ </td>
        </tr>
</table>

In [32]:
import numpy as np

def f(x):
    return (2 * x - 1) / (2 * x + 1)

def int_rectangle_left(f, a, b, step = 0.1):
    X = np.linspace(a, b, int((b - a) / step) + 1)
    return step * np.sum(f(X)[:len(X) - 1])
    
def int_rectangle_right(f, a, b, step = 0.1):
    X = np.linspace(a, b, int((b - a) / step) + 1)
    return step * np.sum(f(X)[1:])

def int_trapezoid(f, a, b, step = 0.1):
    X = np.linspace(a, b, int((b - a) / step) + 1)
    f = f(X)
    return step / 2 * (f[0] + 2 * np.sum(f[1 : len(X) - 1]) + f[-1])

def int_Simpson(f, a, b, step = 0.1):
    X = np.linspace(a, b, 2 * int((b - a) / step) + 1)
    N = len(X) - 1
    f = f(X)
    return step / 6 * (f[0] + 2 * np.sum(f[2 : N : 2])  + 4 * np.sum(f[1 : N : 2]) + f[N])
    
def calc_int(int_method, f, a, b, start_step = 0.1, prec = 0.00001):
    step = start_step
    res = int_method(f, a, b, step)
    c = 0
    while abs(res - int_method(f, a, b, step / 10)) >= prec:
        step /= 10
        res = int_method(f, a, b, step)
        c += 1
    return res, c
    
if __name__ == '__main__':
    np.random.seed(1032192939)
    n = np.random.randint(10)
    print('Variant:', n)
    print(calc_int(int_rectangle_left, f, 0, 2))
    print(calc_int(int_rectangle_right, f, 0, 2))
    print(calc_int(int_trapezoid, f, 0, 2))
    print(calc_int(int_Simpson, f, 0, 2))

Variant: 9
(0.3905521347234617, 4)
(0.39056813472346164, 4)
(0.3905617675660328, 2)
(0.3905588356584965, 0)


$\int\limits_0^2 {\frac{{2x - 1}}{{2x + 1}}dx} = [t = 2x + 1] = {\frac{1}{2}}\int\limits_1^5 {\frac{{t - 2}}{{t}}dt} = {\frac{1}{2}}\int\limits_1^5 {(1 - \frac{2}{t})dt} = {\frac{1}{2}}(\int\limits_1^5 {1dt} + 2\int\limits_1^5 {\frac{1}{t}}) = {\frac{1}{2}}(t - 2\ln(|t|))\Biggr|_{1}^{5} = {\frac{1}{2}}(t - 2\ln(t))\Biggr|_{1}^{5} = {\frac{1}{2}}([5 - 2\ln(5)] - [1 - 2\ln(1)]) = {\frac{1}{2}}(4 - 2\ln(5)) = 2 - \ln(5) \approx 0.390562$

<p><b>Метод прямоугольника по левой стороне:</b> I = 0.3905521347234617, n = 4</p>
<p><b>Метод прямоугольника по правой стороне:</b> I = 0.39056813472346164, n = 4</p>
<p><b>Метод трапеции:</b> I = 0.3905617675660328, n = 2</p>
<p><b>Метод симпсона:</b> I = 0.3905588356584965, n = 0</p>
<p>I - значение определенного итеграла, n - количество итераций для достижения указанной чтоности при заданных параметрах</p>

Как видно, метод трапеции дает более точный результат. Однако, метод Симпсона требует меньше шагов разбиения.

Метод Симсона > метод трапецмм > метод прямоугольника по критерию отношения точности и энергоэффективности.